cocotb/scoreboard.py

```diff
@@ -128,19 +128,18 @@ def compare(self, got, exp, log, strict_type=True):
                                   "Set strict_type=False to avoid this.")
             return
         # Or convert to a string before comparison
-        elif not strict_type:
-            got, exp = str(got), str(exp)
+        if not strict_type:
+            no_match = str(got) != str(exp)
+        else:
+            no_match = got != exp
 
         # Compare directly
-        if got != exp:
+        if no_match:
             self.errors += 1
 
-            # Try our best to print out something useful
-            strgot, strexp = str(got), str(exp)
-
             log.error("Received transaction differed from expected output")
             if not strict_type:
-                log.info("Expected:\n" + hexdump(strexp))
+                log.info("Expected:\n" + hexdump(exp))
             else:
                 log.info("Expected:\n" + repr(exp))
             if not isinstance(exp, str):
```

Comment:
This part here is quite weird as well, it should probably be removed. It tries to iterate over things that are not strings in "words" and printing them. This results in:

```
 90000.00ns ERROR    ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:140  in compare                         Received transaction differed from expected output
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:144  in compare                         Expected:
                                                                                                                     b'\xde\xad\xbe\xef\xde\xad\xbe\xef\xde\xad\xbe\xef\xde\xad\xbe\xef\xde\xad\xbe\xef\xde\xad\xbe\xef\xde\xad\xbe\xef\xde\xad\xbe\xef'
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:148  in compare                         222
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:148  in compare                         173
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:148  in compare                         190
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:148  in compare                         239
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:148  in compare                         222
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:148  in compare                         173
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:148  in compare                         190
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:148  in compare                         239
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:148  in compare                         222
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:148  in compare                         173
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:148  in compare                         190
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:148  in compare                         239
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:148  in compare                         222
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:148  in compare                         173
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:148  in compare                         190
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:148  in compare                         239
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:148  in compare                         222
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:148  in compare                         173
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:148  in compare                         190
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:148  in compare                         239
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:148  in compare                         222
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:148  in compare                         173
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:148  in compare                         190
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:148  in compare                         239
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:148  in compare                         222
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:148  in compare                         173
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:148  in compare                         190
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:148  in compare                         239
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:148  in compare                         222
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:148  in compare                         173
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:148  in compare                         190
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:148  in compare                         239
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:154  in compare                         Received:
                                                                                                                     b'\xde\xad\xbe\xef'
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:158  in compare                         222
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:158  in compare                         173
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:158  in compare                         190
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:158  in compare                         239
```

Reply:
```patch
diff --git a/cocotb/scoreboard.py b/cocotb/scoreboard.py
index f8a23d24..36d664d4 100755
--- a/cocotb/scoreboard.py
+++ b/cocotb/scoreboard.py
@@ -142,22 +142,10 @@ class Scoreboard:
                 log.info("Expected:\n" + hexdump(exp))
             else:
                 log.info("Expected:\n" + repr(exp))
-            if not isinstance(exp, str):
-                try:
-                    for word in exp:
-                        log.info(str(word))
-                except Exception:
-                    pass
             if not strict_type:
                 log.info("Received:\n" + hexdump(got))
             else:
                 log.info("Received:\n" + repr(got))
-            if not isinstance(got, str):
-                try:
-                    for word in got:
-                        log.info(str(word))
-                except Exception:
-                    pass
             log.warning("Difference:\n%s" % hexdiffs(exp, got))
             if self._imm:
                 raise TestFailure("Received transaction differed from expected "
```

That makes the output 100% nice in my use case.

```
 90000.00ns ERROR    ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:140  in compare                         Received transaction differed from expected output
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:144  in compare                         Expected:
                                                                                                                     b'\xde\xad\xbe\xef\xde\xad\xbe\xef\xde\xad\xbe\xef\xde\xad\xbe\xef\xde\xad\xbe\xef\xde\xad\xbe\xef\xde\xad\xbe\xef\xde\xad\xbe\xef'
 90000.00ns INFO     ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:148  in compare                         Received:
                                                                                                                     b'\xde\xad\xbe\xef'
 90000.00ns WARNING  ..scoreboard.fejkon_fc_debug.st_out       scoreboard.py:149  in compare                         Difference:
                                                                                                                     0000      DEADBEEFDEADBEEF DEADBEEFDEADBEEF ................
                                                                                                                          0000 DEADBEEF                          ....            
                                                                                                                     0004      DEADBEEFDEADBEEF DEADBEEFDEADBEEF ................
                                                                                                                     
 90000.00ns INFO     ..Pkts._monitor_recv.0x7f624ba9dc18        scheduler.py:503  in unschedule                      Test stopped by this forked coroutine

```

Reply:
I think the intent here was to give (only slightly) helpful outputs for lists, but given that blows up on the `hexdiffs` line anyway, I'd say we don't care. Your patch looks fine to me.

Reply:
Alternatively, we could change `isinstance(exp, str)` to `isinstance(exp, bytes)`.

Reply:
Yeah, that would work fine too - either way, not much of the scoreboard makes sense for non-bytes types anyway.